### PR TITLE
11.0 mig stock picking package preparation

### PR DESCRIPTION
**Pull request by SimoRubi**
Standard migration

### CODE REVIEW

**Comment by SimoRubi** [comment]
Tests still fail due to https://github.com/odoo/odoo/pull/26850

**Comment by SimoRubi** [comment]
Now that https://github.com/odoo/odoo/pull/26850 is merged, @OCA/logistics-maintainers could you please review?
Thanks

File: stock_picking_package_preparation/__init__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Author: Guewen Baconnier
```
Comment (nikul-serpentcs) [inline stock_picking_package_preparation/__init__.py]: @SimoRubi Remove Copyright from all  `__init__.py` file.

File: stock_picking_package_preparation/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+              'Odoo Community Association (OCA)',
+    'maintainer': 'Camptocamp',
+    'license': 'AGPL-3',
+    'category': 'Warehouse Management',
+    'depends': ['stock'],
+    'website': 'https://github.com/OCA/stock-logistics-workflow/tree/'
+               '11.0/stock_picking_package_preparation',
+    'data': [
+        'views/stock_picking_package_preparation_view.xml',
+        'security/ir.model.access.csv',
```
Comment (nikul-serpentcs) [inline stock_picking_package_preparation/__manifest__.py:17]: @SimoRubi IMP `security`  file always first.

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

File: stock_picking_package_preparation/__init__.py (hunk 1 above)
Comment (SimoRubi) [reply stock_picking_package_preparation/__init__.py]: Thank for the review @nikul-serpentcs, done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/d922d5eb4af6d7d3d6cd338b8cc08ac180f3cb34

File: stock_picking_package_preparation/__manifest__.py (hunk 2 above)
Comment (SimoRubi) [reply stock_picking_package_preparation/__manifest__.py:17]: Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/d922d5eb4af6d7d3d6cd338b8cc08ac180f3cb34

**Comment by SimoRubi** [comment]
@nikul-serpentcs is this ok for you now?

**Comment by SimoRubi** [comment]
@nikul-serpentcs if this is ok for you, please do the review github-style :) thanks

**Comment by nikul-serpentcs** [comment]
@SimoRubi Please attend `codecov` also, It's still False

File: stock_picking_package_preparation/README.rst [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+.. contents::
+   :local:
+
+Usage
+=====
+
+To use this module, you need to:
+
+ * Go to Warehouse > Package Preparation
```
Comment (tafaRU) [inline stock_picking_package_preparation/README.rst]: s/Warehouse/Inventory

**Review by tafaRU — CHANGES_REQUESTED** [review]
On runbot I created `PACK0000002`. Now if I click on the button *Cancel* I get the following error:

```
The operation cannot be completed, probably due to the following:
- deletion: you may be trying to delete a record while other records still reference it
- creation/update: a mandatory field is not correctly set

[object with reference: Packing Operation - stock.move.line]
```
Could you please check it?

File: stock_picking_package_preparation/readme/USAGE.rst [hunk 4]
```diff
@@ -0,0 +1,14 @@
+To use this module, you need to:
+
+ * Go to Warehouse > Package Preparation
+
+From there, you can create a new preparation.
+Choose a partner then the selection of pickings you want in the package.
+You can choose a logistic unit or a packaging, they will be set on the
```
Comment (tafaRU) [inline stock_picking_package_preparation/readme/USAGE.rst]: How can I choose a logistic unit on the pack?

**Review by eLBati — CHANGES_REQUESTED** [review]
Tested steps:

 - Create a delivery order, add a new line and set quantity done = 1
 - Create a package preparation with the previously created picking
 - Click put in pack

System sets picking line quantity done to 0

File: stock_picking_package_preparation/models/stock_picking_package_preparation.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
+# Copyright 2015 Guewen Baconnier
+# Copyright 2016 Lorenzo Battistini - Agile Business Group
+# Copyright 2016 Alessio Gerace - Agile Business Group
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api, _
+from odoo.exceptions import Warning as UserError
```
Comment (rousseldenis) [inline stock_picking_package_preparation/models/stock_picking_package_preparation.py]: You can import UserError directly

File: stock_picking_package_preparation/README.rst (hunk 3 above)
Comment (SimoRubi) [reply stock_picking_package_preparation/README.rst]: Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/3b10189d10ac3fa3e82aeeca990a61c6b566afbd

File: stock_picking_package_preparation/readme/USAGE.rst (hunk 4 above)
Comment (SimoRubi) [reply stock_picking_package_preparation/readme/USAGE.rst]: Logistic units existed in v8, as far as I know they were deleted in v10 already, anyway I've deleted that line in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/3b10189d10ac3fa3e82aeeca990a61c6b566afbd

**Comment by SimoRubi** [comment]
@tafaRU I fixed the error you described in https://github.com/OCA/stock-logistics-workflow/pull/448#pullrequestreview-159416061 with https://github.com/OCA/stock-logistics-workflow/pull/448/commits/3b10189d10ac3fa3e82aeeca990a61c6b566afbd

File: stock_picking_package_preparation/models/stock_picking_package_preparation.py [hunk 6]
```diff
@@ -201,7 +201,9 @@ def _generate_pack(self):
                 picking.action_assign()
             move_lines |= picking.move_line_ids
 
-        for move_line in move_lines:
+        # If the stock.move.line has no processed quantity,
+        # set it equal to the initial demand
+        for move_line in move_lines.filtered(lambda ml: ml.qty_done == 0):
```
Comment (rousseldenis) [inline stock_picking_package_preparation/models/stock_picking_package_preparation.py]: @SimoRubi Careful. On fields with precision, use float_compare()

**Comment by SimoRubi** [comment]
@eLBati I fixed your use-case in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/814764172396643c1004e1ac67a61fed5f03a04d and added the test `test_put_in_pack_keep_done_qty` to avoid the error to happen again.

The confusion is generated, in my opinion, from the different form presented when creating a picking in v10 vs v11:

- In v10 the only field that you were able to see was an anonymous *Quantity* (`product_uom_qty`)
- In v11 you can see both *Initial Demand* (`product_uom_qty`) and *Done quantity* (`quantity_done`)

Then setting only the *Done quantity* as you described in your use-case was not possible in v10 that's why that use-case was not covered **yet**

File: stock_picking_package_preparation/models/stock_picking_package_preparation.py (hunk 5 above)
Comment (SimoRubi) [reply stock_picking_package_preparation/models/stock_picking_package_preparation.py]: Thanks for the review @rousseldenis, I missed that.
Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/169ab09cecbf3762a29dcb670bf29971fe120209

File: stock_picking_package_preparation/models/stock_picking_package_preparation.py (hunk 6 above)
Comment (SimoRubi) [reply stock_picking_package_preparation/models/stock_picking_package_preparation.py]: Thanks, I didn't thought of that! Looking around a bit I found `float_is_zero` that is just what I need :) Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/169ab09cecbf3762a29dcb670bf29971fe120209

**Comment by SimoRubi** [comment]
@nikul-serpentcs the coverage is already above 85% and I think I won't add more tests.
If you find use-cases that result in an error please tell me or add the tests you want creating a PR to my repository.

**Comment by SimoRubi** [comment]
Thanks for your review @tafaRU, I have fixed [what you noticed](https://github.com/OCA/stock-logistics-workflow/pull/448#commitcomment-30685655) in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/18a90f1047417d727c615489403cf32a2a56afc1.
Is this ok for you now?

**Comment by rousseldenis** [comment]
> @eLBati I fixed your use-case in [8147641](https://github.com/OCA/stock-logistics-workflow/commit/814764172396643c1004e1ac67a61fed5f03a04d) and added the test `test_put_in_pack_keep_done_qty` to avoid the error to happen again.
> 
> The confusion is generated, in my opinion, from the different form presented when creating a picking in v10 vs v11:
> 
> * In v10 the only field that you were able to see was an anonymous _Quantity_ (`product_uom_qty`)
> * In v11 you can see both _Initial Demand_ (`product_uom_qty`) and _Done quantity_ (`quantity_done`)
> 
> Then setting only the _Done quantity_ as you described in your use-case was not possible in v10 that's why that use-case was not covered **yet**

@SimoRubi In v10, the done quantity is present in pack operations.

File: stock_picking_package_preparation/models/stock_picking_package_preparation.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+            preparation.quant_ids = package._get_contained_quants()
+
+    @api.multi
+    @api.depends('package_id',
+                 'package_id.quant_ids')
+    def _compute_weight(self):
+        for preparation in self:
+            package = preparation.package_id
+            if not package:
+                return
```
Comment (rousseldenis) [inline stock_picking_package_preparation/models/stock_picking_package_preparation.py]: @SimoRubi I think this should be 'continue' instead as self can be a multi recordset

Comment (SimoRubi) [reply stock_picking_package_preparation/models/stock_picking_package_preparation.py]: Thanks @rousseldenis, I missed that. Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/20954520e6e2decb1728cada9ea7c7b14f583ccb

**Comment by SimoRubi** [comment]
> > @eLBati I fixed your use-case in [8147641](https://github.com/OCA/stock-logistics-workflow/commit/814764172396643c1004e1ac67a61fed5f03a04d) and added the test `test_put_in_pack_keep_done_qty` to avoid the error to happen again.
> > The confusion is generated, in my opinion, from the different form presented when creating a picking in v10 vs v11:
> > 
> > * In v10 the only field that you were able to see was an anonymous _Quantity_ (`product_uom_qty`)
> > * In v11 you can see both _Initial Demand_ (`product_uom_qty`) and _Done quantity_ (`quantity_done`)
> > 
> > Then setting only the _Done quantity_ as you described in your use-case was not possible in v10 that's why that use-case was not covered **yet**
> 
> @SimoRubi In v10, the done quantity is present in pack operations.

@rousseldenis during the creation of a picking (it is still in draft state), in v10 only the tab **Initial Demand** is shown, and looks like there is no `quantity_done` (screenshot from runbot):
image
Are you suggesting that its visibility is due to some configuration? Where can you see that field? Thanks

**Review by eLBati — APPROVED** [review]
Reviewed and tested

**Comment by GSLabIt** [comment]
Any eta to be merged?

**Comment by rousseldenis** [comment]
@tafaRU could you update your review?

**Review by rousseldenis — APPROVED** [review]
Code review

**Review by tafaRU — APPROVED** [review]

**Comment by SimoRubi** [comment]
@joezsweet now that this is merged, I'll start working on https://github.com/OCA/stock-logistics-workflow/pull/447 but I can't give you an ETA as it depends on the OCA reviews.

**Pull request merged**